Add Google Cloud Spanner binding.

scode opened the pull request:
This adds a (basic) Google Cloud Spanner binding. Caveats:

* Only application default credentials are supported.
* Primary key values are not properly escaped during reads because
  I could not find the necessary information to do so with Spanner's
  SQL dialect (probably my mistake).

scode commented:
Note that I've never worked on the code base before and this was a "minimal until it worked for me" kind of effort. Some inspiration taken from other bindings. Please let me know if I'm not following best practices etc.

scode commented on googlespanner/pom.xml:
```diff
@@ -0,0 +1,68 @@
+      <artifactId>guava</artifactId>
+      <version>21.0</version>
+    </dependency>
+    <!-- Additionally, avoid a conflicting older -jdk15 variant of Guava ending up on the classpath. -->
```
Note that the `-jdk15` variant is still pulled in for other bindings. However when running with a binding tailored classpath as done by:

```
./bin/ycsb run googlespanner -P wload -threads 20 -target 200
```

This does not cause a problem. However, if uberjars are generated or runs are otherwise done in such a way as to contain a global set of dependencies, it would still fail.

The failures happen at runtime due to a missing method in one of the Guava classes.

risdenk commented on googlespanner/pom.xml:
```diff
@@ -0,0 +1,68 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+Copyright (c) 2016 YCSB contributors. All rights reserved.
```
Can this be updated to 2017?

risdenk commented on googlespanner/README.md:
```diff
@@ -0,0 +1,19 @@
+<!--
+Copyright (c) 2016 YCSB contributors. All rights reserved.
```
2017

risdenk commented on googlespanner/README.md:
```diff
@@ -0,0 +1,19 @@
+LICENSE file.
+-->
+
+# Google Spanner Driver for YCSB
```
Looks like this is empty? It would be great to follow one of the other bindings and talk about how to use.

scode replied:
Will do, sorry about that.

risdenk commented on googlespanner/src/main/java/com/yahoo/ycsb/db/GoogleSpannerClient.java:
```diff
@@ -0,0 +1,185 @@
+ * LICENSE file.
+ */
+
+package com.yahoo.ycsb.db;
```
This is something new that we are trying to do. Can you make sure that this is in its own package? com.yahoo.ycsb.db.googlespanner or something similar?

This will be a change in most of the files.

scode replied:
Sure, that's my preference too. I just aped what another binding was doing and wasn't sure what the preference was.

risdenk commented on googlespanner/src/main/java/com/yahoo/ycsb/db/GoogleSpannerClient.java:
```diff
@@ -0,0 +1,185 @@
+  @Override
+  public Status read(String table, String key, Set<String> fields,
+          HashMap<String, ByteIterator> result) {
+    // XXX: Key is not escaped. Unclear to me at this time how to do this in Spanner's SQL dialect.
```
What happens in this case? Errors?

scode replied:
Standard escaping problem. Depending on whether or not we want to rely on the workload generating keys that can be naively put between single quotes, this code is wrong and arbitrary SQL may be executed. Whether or not this is acceptable given the intentions of YCSB workload generation I'm not sure, but I timed out on attempting to find out what escaping is required in the Spanner dialect.

scode commented:
@risdenk There is an SQL language for *queries*, but Spanner doesn't allow updates through SQL. I haven't looked at https://cloud.google.com/spanner/docs/partners/drivers but unless it's implementing rewriting of SQL itself I don't expect it to provide that. The mechanism used here to talk to Spanner is that which is documented in the Spanner documentation.

risdenk commented:
@scode Thanks for the SQL explanation.

scode commented:
@risdenk Addressed feedback.

I do want to call attention to the comment in the code about conflicting dependencies. I don't know if this solution is acceptable since I don't expect it to run in a context where all sub modules dependencies have been put on the class path.

The dependency exclusion I added to the `googlespanner` submodule here can probably be duplicated on other modules, provided that we don't care about JDK 1.5 support (the conflicting dependency being a `-jdk5` version of Guava). Since Guava is extremely high quality and release engineered, I don't expect backwards compatibility concerns being very likely - but they are possible and I'm not set up to test other Guava consumers that may be affected by such a change.

risdenk commented on googlespanner/README.md:
```diff
@@ -0,0 +1,63 @@
+* Only Application Default Credentials are supported. In other words,
+  if running on a Google Cloud instance with permissions granted when
+  creating it, authentication will seamlessly work. But use of
+  explicit keys is not implemented.
```
explicit keys being like access key? This is slightly confusing with the next line talking about primary keys for inserting/reading.

scode replied:
Adjusted to hopefully be clearer, and link to documentation.

Yes, like an AWS access key.

siamaktz commented:
Thank you for contributing this, we actually have an official YCSB client for Cloud Spanner by Google that we'd like to merge. Would it be okay to drop this in favor of the official one? Thanks.

scode commented:
Sure
